src/position.cpp

```diff
@@ -563,16 +563,23 @@ bool Position::legal(Move m) const {
             && !(attacks_bb<BISHOP>(ksq, occupied) & pieces(~us, QUEEN, BISHOP));
   }
 
-  // If the moving piece is a king, check whether the destination
-  // square is attacked by the opponent. Castling moves are checked
-  // for legality during move generation.
+  // In case of Chess960, verify that when moving the castling rook we do not
+  // discover some hidden checker.
+  // For instance an enemy queen in SQ_A1 when castling rook is in SQ_B1.
+  if (type_of(m) == CASTLING)
+      return   !chess960
+            || (  to = relative_square(us, to > from ? SQ_G1 : SQ_C1),
+                !(attacks_bb<ROOK>(to, pieces() ^ from) & pieces(~us, ROOK, QUEEN)));
```

Comment:
So now the `to` square is correct but `from` should be `rfrom` like it was in the original. i.e. `to_sq(m)`. Trying to cram all this inside the if-condition is getting too tricky in my opinion.